AssemblerLib/VectorMatrixAssembler.h

```diff
@@ -49,15 +57,41 @@ class VectorMatrixAssembler
     {
         assert(_data_pos.size() > id);
 
-        LocalToGlobalIndexMap::RowColumnIndices const& indices = _data_pos[id];
+        std::vector<GlobalIndexType> indices;
+
+        // Local matrices and vectors will always be ordered by component,
+        // no matter what the order of the global matrix is.
+        for (unsigned c=0; c<_data_pos.getNumComponents(); ++c)
+        {
+            auto const& idcs = _data_pos(id, c).rows;
+            indices.reserve(indices.size() + idcs.size());
+            indices.insert(indices.end(), idcs.begin(), idcs.end());
+        }
+
+        std::vector<double> localX;
+        std::vector<double> localX_pts;
+
+        if (_x)         localX.reserve(indices.size());
+        if (_x_prev_ts) localX_pts.reserve(indices.size());
+
+        for (auto i : indices)
+        {
+            if (_x)         localX.emplace_back(_x->get(i));
```

Comment:
> Get values entry by entry is slow in PETSc.

Is there really a big performance impact from getting values index-by-index? If I use `get(const std::vector<PetscInt> &e_idxs,  T_SUBVEC &sub_vec)`, the PETSc function probably is faster (though, not with zero runtime), but results have to be cached.

Additionally, extending all vector classes would clutter this PR. Sure, this is a rather weak argument. But, as @endJunction and you pointed out, there is an issue with negative indices and PETSc. This is definitely out of scope of this PR. I think ghost nodes should be handled transparently in the class `PetscVector`.

I suggest to leave this issue open for this PR and to work on a solution for it afterwards.

Do you agree?